File: android/src/main/java/com/tencent/chat/tencent_cloud_chat_sdk/manager/CommunityManager.java
package com.tencent.chat.tencent_cloud_chat_sdk.manager;

import com.tencent.chat.tencent_cloud_chat_sdk.util.CommonUtil;
import com.tencent.imsdk.v2.V2TIMCallback;
import com.tencent.imsdk.v2.V2TIMCommunityListener;
import com.tencent.imsdk.v2.V2TIMCreateGroupMemberInfo;
import com.tencent.imsdk.v2.V2TIMGroupInfo;
import com.tencent.imsdk.v2.V2TIMManager;
import com.tencent.imsdk.v2.V2TIMPermissionGroupInfo;
import com.tencent.imsdk.v2.V2TIMPermissionGroupInfoResult;
import com.tencent.imsdk.v2.V2TIMPermissionGroupMemberInfoResult;
import com.tencent.imsdk.v2.V2TIMPermissionGroupMemberOperationResult;
import com.tencent.imsdk.v2.V2TIMPermissionGroupOperationResult;
import com.tencent.imsdk.v2.V2TIMTopicInfo;
import com.tencent.imsdk.v2.V2TIMTopicOperationResult;
import com.tencent.imsdk.v2.V2TIMTopicPermissionResult;
import com.tencent.imsdk.v2.V2TIMValueCallback;

import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class CommunityManager {
    static List<MethodChannel> channels = new LinkedList<>();
    public CommunityManager(MethodChannel channel){
        channels.add(channel);
    }
    static V2TIMCommunityListener listener = new V2TIMCommunityListener() {
        @Override
        public void onCreateTopic(String groupID, String topicID) {
            super.onCreateTopic(groupID, topicID);
            HashMap<String,Object> data = new HashMap<>();
            data.put("groupID",groupID);
            data.put("topicID",topicID);
            makeaddCommunityListenerEventData("onCreateTopic",data,"");
        }

        @Override
        public void onDeleteTopic(String groupID, List<String> topicIDList) {

            super.onDeleteTopic(groupID, topicIDList);
            HashMap<String,Object> data = new HashMap<>();
            data.put("groupID",groupID);
            data.put("topicIDList",topicIDList);
            makeaddCommunityListenerEventData("onDeleteTopic",data,"");
        }

        @Override
        public void onChangeTopicInfo(String groupID, V2TIMTopicInfo topicInfo) {

            super.onChangeTopicInfo(groupID, topicInfo);
            HashMap<String,Object> data = new HashMap<>();
            data.put("groupID",groupID);
            data.put("topicInfo",CommonUtil.convertV2TIMTopicInfoToMap(topicInfo));
            makeaddCommunityListenerEventData("onChangeTopicInfo",data,"");
        }

        @Override
        public void onReceiveTopicRESTCustomData(String topicID, byte[] customData) {
            super.onReceiveTopicRESTCustomData(topicID, customData);
            HashMap<String,Object> data = new HashMap<>();
            data.put("topicID",topicID);
            data.put("customData",new String(customData));
            makeaddCommunityListenerEventData("onReceiveTopicRESTCustomData",data,"");
        }

        @Override
        public void onCreatePermissionGroup(String groupID, V2TIMPermissionGroupInfo permissionGroupInfo) {
            super.onCreatePermissionGroup(groupID, permissionGroupInfo);
            HashMap<String,Object> data = new HashMap<>();
            data.put("groupID",groupID);
            data.put("permissionGroupInfo",CommonUtil.convertV2TIMPermissionGroupInfoToMap(permissionGroupInfo));
            makeaddCommunityListenerEventData("onCreatePermissionGroup",data,"");
        }

        @Override
        public void onDeletePermissionGroup(String groupID, List<String> permissionGroupIDList) {
            super.onDeletePermissionGroup(groupID, permissionGroupIDList);
            HashMap<String,Object> data = new HashMap<>();
            data.put("groupID",groupID);
            data.put("permissionGroupIDList",permissionGroupIDList);
            makeaddCommunityListenerEventData("onDeletePermissionGroup",data,"");
        }

        @Override
        public void onChangePermissionGroupInfo(String groupID, V2TIMPermissionGroupInfo permissionGroupInfo) {
            super.onChangePermissionGroupInfo(groupID, permissionGroupInfo);
            HashMap<String,Object> data = new HashMap<>();
            data.put("groupID",groupID);
            data.put("permissionGroupInfo",CommonUtil.convertV2TIMPermissionGroupInfoToMap(permissionGroupInfo));
            makeaddCommunityListenerEventData("onChangePermissionGroupInfo",data,"");
        }

        @Override
        public void onAddMembersToPermissionGroup(String groupID, String permissionGroupID, List<String> memberIDList) {
            super.onAddMembersToPermissionGroup(groupID, permissionGroupID, memberIDList);
            HashMap<String,Object> data = new HashMap<>();
            data.put("groupID",groupID);
            data.put("permissionGroupID",permissionGroupID);
            data.put("memberIDList",memberIDList);
            makeaddCommunityListenerEventData("onAddMembersToPermissionGroup",data,"");
        }

        @Override
        public void onRemoveMembersFromPermissionGroup(String groupID, String permissionGroupID, List<String> memberIDList) {
            super.onRemoveMembersFromPermissionGroup(groupID, permissionGroupID, memberIDList);
            HashMap<String,Object> data = new HashMap<>();
            data.put("groupID",groupID);
            data.put("permissionGroupID",permissionGroupID);
            data.put("memberIDList",memberIDList);
            makeaddCommunityListenerEventData("onRemoveMembersFromPermissionGroup",data,"");
        }

        @Override
        public void onAddTopicPermission(String groupID, String permissionGroupID, HashMap<String, Long> topicPermissionMap) {
            super.onAddTopicPermission(groupID, permissionGroupID, topicPermissionMap);
            HashMap<String,Object> data = new HashMap<>();
            data.put("groupID",groupID);
            data.put("permissionGroupID",permissionGroupID);
            data.put("topicPermissionMap",topicPermissionMap);
            makeaddCommunityListenerEventData("onAddTopicPermission",data,"");
        }

        @Override
        public void onDeleteTopicPermission(String groupID, String permissionGroupID, List<String> topicIDList) {
            super.onDeleteTopicPermission(groupID, permissionGroupID, topicIDList);
            HashMap<String,Object> data = new HashMap<>();
            data.put("groupID",groupID);
            data.put("permissionGroupID",permissionGroupID);
            data.put("topicIDList",topicIDList);
            makeaddCommunityListenerEventData("onDeleteTopicPermission",data,"");
        }

        @Override
        public void onModifyTopicPermission(String groupID, String permissionGroupID, HashMap<String, Long> topicPermissionMap) {
            super.onModifyTopicPermission(groupID, permissionGroupID, topicPermissionMap);
            HashMap<String,Object> data = new HashMap<>();
            data.put("groupID",groupID);
            data.put("permissionGroupID",permissionGroupID);
            data.put("topicPermissionMap",topicPermissionMap);
            makeaddCommunityListenerEventData("onModifyTopicPermission",data,"");
        }
    };
    private static  <T> void  makeaddCommunityListenerEventData(String type,T data, String listenerUuid){
        for (MethodChannel channel : channels) {
            CommonUtil.emitEvent(channel,"communityListener",type,data, listenerUuid);
        }
    }
    public void addCommunityListener(MethodCall methodCall, final MethodChannel.Result result){
        V2TIMManager.getCommunityManager().addCommunityListener(listener);
        CommonUtil.returnSuccess(result,"");
    }
    public void removeCommunityListener(MethodCall methodCall, final MethodChannel.Result result){
        V2TIMManager.getCommunityManager().removeCommunityListener(listener);
        CommonUtil.returnSuccess(result,"");
    }

    public void createCommunity(MethodCall methodCall, final MethodChannel.Result result){
        HashMap<String,Object> info = methodCall.argument("info");
        LinkedList<HashMap<String,Object>> memberList = methodCall.argument("memberList");

        LinkedList<V2TIMCreateGroupMemberInfo> n_member_info = new LinkedList();
        for (HashMap<String, Object> stringObjectHashMap : memberList) {
            V2TIMCreateGroupMemberInfo role = new V2TIMCreateGroupMemberInfo();
            if(stringObjectHashMap.containsKey("role")){
                role.setRole((Integer) stringObjectHashMap.get("role"));
            }
            if(stringObjectHashMap.containsKey("userID")){
                role.setUserID((String) stringObjectHashMap.get("userID"));
            }
            n_member_info.add(role);
        }
        V2TIMManager.getCommunityManager().createCommunity(CommonUtil.HashMapToV2TIMGroupInfo(info), n_member_info, new V2TIMValueCallback<String>() {
            @Override
            public void onSuccess(String s) {
                CommonUtil.returnSuccess(result,s);
            }

            @Override
            public void onError(int code, String desc) {
                CommonUtil.returnError(result,code,desc);
            }
        });
    }
    public void createPermissionGroupInCommunity (MethodCall methodCall, final MethodChannel.Result result){
        HashMap<String,Object> info = methodCall.argument("info");
        V2TIMManager.getCommunityManager().createPermissionGroupInCommunity(CommonUtil.HashMapToV2TIMPermissionGroupInfo(info), new V2TIMValueCallback<String>() {
            @Override
            public void onSuccess(String s) {
                CommonUtil.returnSuccess(result,s);
            }

            @Override
            public void onError(int code, String desc) {
                CommonUtil.returnError(result,code,desc);
            }
        });
    }
    public void deletePermissionGroupFromCommunity(MethodCall methodCall, final MethodChannel.Result result){
        String groupID =  methodCall.argument("groupID");
        List<String> permissionGroupIDList = methodCall.argument("permissionGroupIDList");
        V2TIMManager.getCommunityManager().deletePermissionGroupFromCommunity(groupID, permissionGroupIDList, new V2TIMValueCallback<List<V2TIMPermissionGroupOperationResult>>() {
            @Override
            public void onSuccess(List<V2TIMPermissionGroupOperationResult> v2TIMPermissionGroupOperationResults) {
                CommonUtil.returnSuccess(result,CommonUtil.convertV2TIMPermissionGroupOperationResultListToMap(v2TIMPermissionGroupOperationResults));
            }

            @Override
            public void onError(int code, String desc) {
                CommonUtil.returnError(result,code,desc);
            }
        });
    }
    public void modifyPermissionGroupInfoInCommunity(MethodCall methodCall, final MethodChannel.Result result){
        HashMap<String,Object> info = methodCall.argument("info");
        V2TIMManager.getCommunityManager().modifyPermissionGroupInfoInCommunity(CommonUtil.HashMapToV2TIMPermissionGroupInfo(info), new V2TIMCallback() {
            @Override
            public void onSuccess() {
                CommonUtil.returnSuccess(result,"");
            }

            @Override
            public void onError(int code, String desc) {
                CommonUtil.returnError(result,code,desc);
            }
        });
    }
    public void getJoinedPermissionGroupListInCommunity(MethodCall methodCall, final MethodChannel.Result result){
        String groupID =  methodCall.argument("groupID");
        V2TIMManager.getCommunityManager().getJoinedPermissionGroupListInCommunity(groupID, new V2TIMValueCallback<List<V2TIMPermissionGroupInfoResult>>() {
            @Override
            public void onSuccess(List<V2TIMPermissionGroupInfoResult> v2TIMPermissionGroupInfoResults) {
                CommonUtil.returnSuccess(result,CommonUtil.convertV2TIMPermissionGroupInfoResultListToMap(v2TIMPermissionGroupInfoResults));
            }

            @Override
            public void onError(int code, String desc) {
                CommonUtil.returnError(result,code,desc);
            }
        });
    }
    public void getPermissionGroupListInCommunity(MethodCall methodCall, final MethodChannel.Result result){
        String groupID =  methodCall.argument("groupID");
        List<String> permissionGroupIDList = methodCall.argument("permissionGroupIDList");
        V2TIMManager.getCommunityManager().getPermissionGroupListInCommunity(groupID, permissionGroupIDList, new V2TIMValueCallback<List<V2TIMPermissionGroupInfoResult>>() {
            @Override
            public void onSuccess(List<V2TIMPermissionGroupInfoResult> v2TIMPermissionGroupInfoResults) {
                CommonUtil.returnSuccess(result,CommonUtil.convertV2TIMPermissionGroupInfoResultListToMap(v2TIMPermissionGroupInfoResults));
            }

            @Override
            public void onError(int code, String desc) {
                CommonUtil.returnError(result,code,desc);
            }
        });
    }
    public void addCommunityMembersToPermissionGroup(MethodCall methodCall, final MethodChannel.Result result){

        String groupID =  methodCall.argument("groupID");
        String permissionGroupID = methodCall.argument("permissionGroupID");
        List<String> memberList = methodCall.argument("memberList");
        V2TIMManager.getCommunityManager().addCommunityMembersToPermissionGroup(groupID, permissionGroupID, memberList, new V2TIMValueCallback<List<V2TIMPermissionGroupMemberOperationResult>>() {
            @Override
            public void onSuccess(List<V2TIMPermissionGroupMemberOperationResult> v2TIMPermissionGroupMemberOperationResults) {

                CommonUtil.returnSuccess(result,CommonUtil.convertV2TIMPermissionGroupMemberOperationResultListToMap(v2TIMPermissionGroupMemberOperationResults));
            }

            @Override
            public void onError(int code, String desc) {
                CommonUtil.returnError(result,code,desc);
            }
        });
    }
    public void removeCommunityMembersFromPermissionGroup(MethodCall methodCall, final MethodChannel.Result result){
        String groupID =  methodCall.argument("groupID");
        String permissionGroupID = methodCall.argument("permissionGroupID");
        List<String> memberList = methodCall.argument("memberList");
        V2TIMManager.getCommunityManager().removeCommunityMembersFromPermissionGroup(groupID, permissionGroupID, memberList, new V2TIMValueCallback<List<V2TIMPermissionGroupMemberOperationResult>>() {
            @Override
            public void onSuccess(List<V2TIMPermissionGroupMemberOperationResult> v2TIMPermissionGroupMemberOperationResults) {

                CommonUtil.returnSuccess(result,CommonUtil.convertV2TIMPermissionGroupMemberOperationResultListToMap(v2TIMPermissionGroupMemberOperationResults));
            }

            @Override
            public void onError(int code, String desc) {
                CommonUtil.returnError(result,code,desc);
            }
        });
    }
    public void getCommunityMemberListInPermissionGroup(MethodCall methodCall, final MethodChannel.Result result){
        String groupID =  methodCall.argument("groupID");
        String permissionGroupID = methodCall.argument("permissionGroupID");
        String nextCursor = methodCall.argument("nextCursor");
        V2TIMManager.getCommunityManager().getCommunityMemberListInPermissionGroup(groupID, permissionGroupID, nextCursor, new V2TIMValueCallback<V2TIMPermissionGroupMemberInfoResult>() {
            @Override
            public void onSuccess(V2TIMPermissionGroupMemberInfoResult v2TIMPermissionGroupMemberInfoResult) {
                CommonUtil.returnSuccess(result,CommonUtil.convertV2TIMPermissionGroupMemberInfoResultToMap(v2TIMPermissionGroupMemberInfoResult));
            }

            @Override
            public void onError(int code, String desc) {
                CommonUtil.returnError(result,code,desc);
            }
        });
    }
    public void addTopicPermissionToPermissionGroup(MethodCall methodCall, final MethodChannel.Result result){
        String groupID =  methodCall.argument("groupID");
        String permissionGroupID = methodCall.argument("permissionGroupID");
        HashMap<String,Number> topicPermissionMap = methodCall.argument("topicPermissionMap");
        HashMap<String,Long> customInfoLong = new HashMap<String,Long>();
        Iterator<String> iterator = topicPermissionMap.keySet().iterator();
        while (iterator.hasNext()) {
            String key = iterator.next();
            customInfoLong.put(key,Long.valueOf(topicPermissionMap.get(key).longValue()));
        }
        V2TIMManager.getCommunityManager().addTopicPermissionToPermissionGroup(groupID, permissionGroupID, customInfoLong, new V2TIMValueCallback<List<V2TIMTopicOperationResult>>() {
            @Override
            public void onSuccess(List<V2TIMTopicOperationResult> v2TIMTopicOperationResults) {
                CommonUtil.returnSuccess(result,CommonUtil.convertV2TIMTopicOperationResultListToMap(v2TIMTopicOperationResults));
            }

            @Override
            public void onError(int code, String desc) {
                CommonUtil.returnError(result,code,desc);
            }
        });
    }
    public void deleteTopicPermissionFromPermissionGroup(MethodCall methodCall, final MethodChannel.Result result){
        String groupID =  methodCall.argument("groupID");
        String permissionGroupID = methodCall.argument("permissionGroupID");
        List<String> topicIDList = methodCall.argument("topicIDList");
        V2TIMManager.getCommunityManager().deleteTopicPermissionFromPermissionGroup(groupID, permissionGroupID, topicIDList, new V2TIMValueCallback<List<V2TIMTopicOperationResult>>() {
            @Override
            public void onSuccess(List<V2TIMTopicOperationResult> v2TIMTopicOperationResults) {
                CommonUtil.returnSuccess(result,CommonUtil.convertV2TIMTopicOperationResultListToMap(v2TIMTopicOperationResults));
            }

            @Override
            public void onError(int code, String desc) {
                CommonUtil.returnError(result,code,desc);
            }
        });
    }
    public void modifyTopicPermissionInPermissionGroup(MethodCall methodCall, final MethodChannel.Result result){
        String groupID =  methodCall.argument("groupID");
        String permissionGroupID = methodCall.argument("permissionGroupID");
        HashMap<String,Number> topicPermissionMap = methodCall.argument("topicPermissionMap");
        HashMap<String,Long> customInfoLong = new HashMap<String,Long>();
        Iterator<String> iterator = topicPermissionMap.keySet().iterator();
        while (iterator.hasNext()) {
            String key = iterator.next();
            customInfoLong.put(key,Long.valueOf(topicPermissionMap.get(key).longValue()));
        }
        V2TIMManager.getCommunityManager().modifyTopicPermissionInPermissionGroup(groupID, permissionGroupID, customInfoLong, new V2TIMValueCallback<List<V2TIMTopicOperationResult>>() {
            @Override
            public void onSuccess(List<V2TIMTopicOperationResult> v2TIMTopicOperationResults) {
                CommonUtil.returnSuccess(result,CommonUtil.convertV2TIMTopicOperationResultListToMap(v2TIMTopicOperationResults));
            }

            @Override
            public void onError(int code, String desc) {
                CommonUtil.returnError(result,code,desc);
            }
        });
    }
    public void getTopicPermissionInPermissionGroup(MethodCall methodCall, final MethodChannel.Result result){
        String groupID =  methodCall.argument("groupID");
        String permissionGroupID = methodCall.argument("permissionGroupID");
        List<String> topicIDList = methodCall.argument("topicIDList");
        V2TIMManager.getCommunityManager().getTopicPermissionInPermissionGroup(groupID, permissionGroupID, topicIDList, new V2TIMValueCallback<List<V2TIMTopicPermissionResult>>() {
            @Override
            public void onSuccess(List<V2TIMTopicPermissionResult> v2TIMTopicPermissionResults) {
                CommonUtil.returnSuccess(result,CommonUtil.convertV2TIMTopicPermissionResultListToMap(v2TIMTopicPermissionResults));
            }

            @Override
            public void onError(int code, String desc) {
                CommonUtil.returnError(result,code,desc);
            }
        });
    }
}

File: macos/Classes/manager/CommunityManager.swift
import ImSDKForMac_Plus
import FlutterMacOS


 class CommunityManager {
    var channel: FlutterMethodChannel
    
    
    
    init(channel: FlutterMethodChannel) {
        self.channel = channel
    }
    
    
    
    func createCommunity(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let info = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "info"}) as?  Dictionary<String, Any>;
        let memberList = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "memberList"}) as?  [Dictionary<String, Any>];
        let gruopInfo = V2GroupInfoEntity.init(dict: info!);
        var n_memberList:[V2TIMCreateGroupMemberInfo] = []
        memberList?.forEach({ data in
            let it = V2TIMCreateGroupMemberInfo();
            it.role = data["role"] as? UInt32 ?? 0
            it.userID = data["userID"] as? String ?? ""
            n_memberList.append(it)
        })
        V2TIMManager.sharedInstance().createCommunity(gruopInfo, memberList: n_memberList) { groupid in
            CommonUtils.resultSuccess(call: call, result: result,data: groupid ?? "")
        } fail: {  code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    
    func createPermissionGroupInCommunity(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let info = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "info"}) as?  Dictionary<String, Any>;
        let n_info = V2TimPermissionGroupInfo.init(dict: info!);
        V2TIMManager.sharedInstance().createPermissionGroup(inCommunity: n_info) { gid in
            CommonUtils.resultSuccess(call: call, result: result,data: gid ?? "")
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    
    func deletePermissionGroupFromCommunity(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let groupID = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "groupID"}) as?  String;
        let permissionGroupIDList = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "permissionGroupIDList"}) as?  [String];
        V2TIMManager.sharedInstance().deletePermissionGroup(fromCommunity: groupID, permissionGroupIDList: permissionGroupIDList) { re_arr in
            CommonUtils.resultSuccess(call: call, result: result,data: V2TimPermissionGroupOperationResult.getListDict(infos: re_arr as! [V2TIMPermissionGroupOperationResult]))
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    
    func modifyPermissionGroupInfoInCommunity(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let info = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "info"}) as?  Dictionary<String, Any>;
        let n_info = V2TimPermissionGroupInfo.init(dict: info!);
        V2TIMManager.sharedInstance().modifyPermissionGroupInfo(inCommunity: n_info) {
            CommonUtils.resultSuccess(call: call, result: result)
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    
    func getJoinedPermissionGroupListInCommunity(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let groupID = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "groupID"}) as?  String;
        V2TIMManager.sharedInstance().getJoinedPermissionGroupList(inCommunity: groupID) { ar_arr in
            CommonUtils.resultSuccess(call: call, result: result,data: V2TimPermissionGroupInfoResult.getListDict(infos: ar_arr as! [V2TIMPermissionGroupInfoResult]))
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    
    func getPermissionGroupListInCommunity(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let groupID = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "groupID"}) as?  String;
        let permissionGroupIDList = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "permissionGroupIDList"}) as?  [String];
        V2TIMManager.sharedInstance().getPermissionGroupList(inCommunity: groupID, permissionGroupIDList: permissionGroupIDList) { ar_arr in
            CommonUtils.resultSuccess(call: call, result: result,data: V2TimPermissionGroupInfoResult.getListDict(infos: ar_arr as! [V2TIMPermissionGroupInfoResult]))
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    
    func addCommunityMembersToPermissionGroup(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let groupID = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "groupID"}) as?  String;
        let permissionGroupID = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "permissionGroupID"}) as?  String;
        let memberList = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "memberList"}) as?  [String];
        V2TIMManager.sharedInstance().addCommunityMembers(toPermissionGroup: groupID, permissionGroupID: permissionGroupID, memberList: memberList) { re_arr in
            CommonUtils.resultSuccess(call: call, result: result,data: V2TimPermissionGroupOperationResult.getListDict(infos: re_arr as! [V2TIMPermissionGroupOperationResult]))
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    
    func removeCommunityMembersFromPermissionGroup(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let groupID = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "groupID"}) as?  String;
        let permissionGroupID = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "permissionGroupID"}) as?  String;
        let memberList = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "memberList"}) as?  [String];
        V2TIMManager.sharedInstance().removeCommunityMembers(fromPermissionGroup: groupID, permissionGroupID: permissionGroupID, memberList: memberList) { re_arr in
            CommonUtils.resultSuccess(call: call, result: result,data: V2TimPermissionGroupOperationResult.getListDict(infos: re_arr as! [V2TIMPermissionGroupOperationResult]))
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    
    func getCommunityMemberListInPermissionGroup(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let groupID = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "groupID"}) as?  String;
        let permissionGroupID = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "permissionGroupID"}) as?  String;
        let nextCursor = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "nextCursor"}) as?  String;
        V2TIMManager.sharedInstance().getCommunityMemberList(inPermissionGroup: groupID, permissionGroupID: permissionGroupID, nextCursor: nextCursor) { next_seq, info_list in
            var res: [String: Any] = [:];
            res["nextCursor"] = next_seq;
            res["memberInfoList"] = V2GroupMemberFullInfoEntity.getListDict(infos: info_list as! [V2TIMGroupMemberFullInfo])
            CommonUtils.resultSuccess(call: call, result: result,data: res)
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    
    func addTopicPermissionToPermissionGroup(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let groupID = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "groupID"}) as?  String;
        let permissionGroupID = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "permissionGroupID"}) as?  String;
        let topicPermissionMap = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "topicPermissionMap"}) as?  [String: NSNumber];
        V2TIMManager.sharedInstance().addTopicPermission(toPermissionGroup: groupID, permissionGroupID: permissionGroupID, topicPermissionMap: topicPermissionMap) { re_arr in
            CommonUtils.resultSuccess(call: call, result: result,data: V2TimTopicOperationResult.getListDict(infos: re_arr as! [V2TIMTopicOperationResult]))
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    
    func deleteTopicPermissionFromPermissionGroup(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let groupID = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "groupID"}) as?  String;
        let permissionGroupID = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "permissionGroupID"}) as?  String;
        let topicIDList = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "topicIDList"}) as?  [String];
        V2TIMManager.sharedInstance().deleteTopicPermission(fromPermissionGroup: groupID, permissionGroupID: permissionGroupID, topicIDList: topicIDList) { re_arr in
            CommonUtils.resultSuccess(call: call, result: result,data: V2TimTopicOperationResult.getListDict(infos: re_arr as! [V2TIMTopicOperationResult]))
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    
    func modifyTopicPermissionInPermissionGroup(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let groupID = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "groupID"}) as?  String;
        let permissionGroupID = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "permissionGroupID"}) as?  String;
        let topicPermissionMap = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "topicPermissionMap"}) as?  [String: NSNumber];
        V2TIMManager.sharedInstance().modifyTopicPermission(inPermissionGroup: groupID, permissionGroupID: permissionGroupID, topicPermissionMap: topicPermissionMap) { re_arr in
            CommonUtils.resultSuccess(call: call, result: result,data: V2TimTopicOperationResult.getListDict(infos: re_arr as! [V2TIMTopicOperationResult]))
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    
    func getTopicPermissionInPermissionGroup(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let groupID = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "groupID"}) as?  String;
        let permissionGroupID = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "permissionGroupID"}) as?  String;
        let topicIDList = (call.arguments as? Dictionary<String, Any>)?.first(where: {$0.key == "topicIDList"}) as?  [String];
        V2TIMManager.sharedInstance().getTopicPermission(inPermissionGroup: groupID, permissionGroupID: permissionGroupID, topicIDList: topicIDList) { re_arr in
            CommonUtils.resultSuccess(call: call, result: result,data: V2TimTopicPermissionResult.getListDict(infos: re_arr as! [V2TIMTopicPermissionResult]))
        } fail: { code, desc in
            CommonUtils.resultFailed(desc: desc, code: code, call: call, result: result)
        }

    }
    
}
